src/rust/bitbox02-rust/src/hww/api/api.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2020 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::vec::Vec;
+
+use super::pb;
+use super::pb::request::Request;
+use super::pb::response::Response;
+use bitbox02::commander::Error;
+
+/// Creates an Error response. Corresponds to commander.c:_report_error().
+fn make_error(err: bitbox02::commander::Error) -> Response {
+    use Error::*;
+    let err = match err {
+        COMMANDER_OK => panic!("can't call this function with COMMANDER_OK"),
+        COMMANDER_ERR_INVALID_INPUT => pb::Error {
+            code: 101,
+            message: "invalid input".into(),
+        },
+        COMMANDER_ERR_MEMORY => pb::Error {
+            code: 102,
+            message: "memory".into(),
+        },
+        COMMANDER_ERR_GENERIC => pb::Error {
+            code: 103,
+            message: "generic error".into(),
+        },
+        COMMANDER_ERR_USER_ABORT => pb::Error {
+            code: 104,
+            message: "aborted by the user".into(),
+        },
+        COMMANDER_ERR_INVALID_STATE => pb::Error {
+            code: 105,
+            message: "can't call this endpoint: wrong state".into(),
+        },
+        COMMANDER_ERR_DISABLED => pb::Error {
+            code: 106,
+            message: "function disabled".into(),
+        },
+        COMMANDER_ERR_DUPLICATE => pb::Error {
+            code: 107,
+            message: "duplicate entry".into(),
+        },
+    };
+    Response::Error(err)
+}
+
+/// Encodes a protobuf Response message.
+fn encode(response: Response) -> Vec<u8> {
+    use prost::Message;
+    let response = pb::Response {
+        response: Some(response),
+    };
+    let mut out = Vec::<u8>::new();
+    response.encode(&mut out).unwrap();
+    out
+}
+
+async fn api_set_device_name(
+    pb::SetDeviceNameRequest { name }: &pb::SetDeviceNameRequest,
+) -> Response {
+    use crate::workflow::confirm;
```

Comment:
I find it a bit misleading that some use imports are at the top and the others are scattered around. Harder to use what is used where. Maybe I'm just not used to this in Rust yet.

Reply:
I guess it's a style preference. I like local imports as I can just move the whole function to a new file and it will work the same way, and you know exactly what import is used by which function(s).

Not a hard rule, I do it case by case. Can discuss offline if we should have a style guide.

Reply:
Like this one? https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/items.md

Reply:
That implies all `use` at the top right?

If that is enforced, I might add more small files instead of slightly bigger files with self contained functions. In this PR, maybe move the fn into `set_device_name.rs`.

Reply:
but why would you split because of where the use declarations are?